### ISSUE TITLE
Peformance Problems

### ISSUE THREAD

**Issue by LekoArts** [TryGhost/gatsby-plugin-advanced-sitemap#188]
## Issue Summary

Gatsby maintainer here 👋 

Two of our cloud customers saw some pretty long `onPostBuild` steps in their builds (one OOMs, the other takes 30mins) and we pinpointed it to `gatsby-plugin-advanced-sitemap` to be the culprit.

I'd like to get some more context on the plugin so that we can solve this problem for them and everyone else using it.

For comparison: On a 10k page ours takes 2sec, yours 10sec. On a 50k page ours takes under 10sec, yours around 300sec.

- We updated our own plugin some months ago and it now outputs multiple files by default if there are enough entries. It also got some other improvements: https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-sitemap - What is the main differentiator between your plugin and our official one now?
- Are the differentiations used a lot? What were the use cases?
- See attached image (50k page). The plugin has a O(n^2) runtime since it loops over every page for every page. Inside serialize this happens: https://github.com/TryGhost/gatsby-plugin-advanced-sitemap/blob/master/src/gatsby-node.js#L209-L271 - what's happening there exactly? What is the purpose of the getNodePath function? https://github.com/TryGhost/gatsby-plugin-advanced-sitemap/blob/master/src/gatsby-node.js#L74-L90

I'm happy with whatever outcome we'd agree to if the end result is that users won't get a speed penality.

So outcomes could be:
- The official plugin is sufficient and yours could be deprecated
- The functionality that causes the slowdown can be removed
- The slowdown can be fixed while keeping functionality

### Attachments

![Clipboard 2021-12-08 at 4 27 16 PM (1)](https://user-images.githubusercontent.com/16143594/129694966-c74dd1fa-b341-46fe-88f4-17789a50895a.png)

You can open this in Chrome Node Profiler: [sitemap-cpu-profile.zip](https://github.com/TryGhost/gatsby-plugin-advanced-sitemap/files/6998484/sitemap-cpu-profile.zip)

### To Reproduce

1. Take https://github.com/gatsbyjs/gatsby/tree/master/benchmarks/query-filters-sort and add the sitemap plugin
2. Run it with > 50k pages
3. See huge increase in `onPostBuild` time

### Technical details:

```
  System:
    OS: Linux 5.4 Ubuntu 20.04.2 LTS (Focal Fossa)
    CPU: (8) x64 AMD Ryzen 9 3900X 12-Core Processor
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 14.16.1 - ~/.nvm/versions/node/v14.16.1/bin/node
    Yarn: 1.22.5 - /usr/bin/yarn
    npm: 6.14.12 - ~/.nvm/versions/node/v14.16.1/bin/npm
  Languages:
    Python: 2.7.18 - /usr/bin/python
  Browsers:
    Chrome: 92.0.4515.107
  npmPackages:
    gatsby: ^3.11.0 => 3.11.1
    gatsby-plugin-advanced-sitemap: ^2.0.0 => 2.0.0
  npmGlobalPackages:
    gatsby-cli: 3.8.0
    gatsby-dev-cli: 3.8.0
```

**Comment by yogeshkotadiya** [comment]
@LekoArts Thanks a lot for finding the root cause 🙌 .

I've updated the problem function and the result is significantly improved.
#178

**Comment by Artemdanilov5** [comment]
what's the status of this problem?

**Comment by yogeshkotadiya** [comment]
Fixed in v2.1.0 : #178
